rust/datafusion/src/physical_plan/repartition.rs

```diff
@@ -0,0 +1,336 @@
+    tx: Arc<Mutex<Vec<Sender<Option<ArrowResult<RecordBatch>>>>>>,
+}
+
+#[async_trait]
+impl ExecutionPlan for RepartitionExec {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    /// Get the schema for this execution plan
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn with_new_children(
+        &self,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        match children.len() {
+            1 => Ok(Arc::new(RepartitionExec::try_new(
+                children[0].clone(),
+                self.partitioning.clone(),
+            )?)),
+            _ => Err(DataFusionError::Internal(
+                "RepartitionExec wrong number of children".to_string(),
+            )),
+        }
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.partitioning.clone()
+    }
+
+    async fn execute(&self, partition: usize) -> Result<SendableRecordBatchStream> {
+        // lock mutexes
+        let mut tx = self.tx.lock().await;
+        let mut rx = self.rx.lock().await;
+
+        let num_input_partitions = self.input.output_partitioning().partition_count();
+        let num_output_partitions = self.partitioning.partition_count();
+
+        // if this is the first partition to be invoked then we need to set up initial state
+        if tx.is_empty() {
+            // create one channel per *output* partition
+            for _ in 0..num_output_partitions {
+                // Note that this operator uses unbounded channels to avoid deadlocks because
+                // the output partitions can be read in any order and this could cause input
+                // partitions to be blocked when sending data to output receivers that are not
+                // being read yet. This may cause high memory usage if the next operator is
+                // reading output partitions in order rather than concurrently. One workaround
+                // for this would be to add spill-to-disk capabilities.
+                let (sender, receiver) = unbounded::<Option<ArrowResult<RecordBatch>>>();
+                tx.push(sender);
+                rx.push(receiver);
+            }
+            // launch one async task per *input* partition
+            for i in 0..num_input_partitions {
+                let input = self.input.clone();
+                let mut tx = tx.clone();
+                let partitioning = self.partitioning.clone();
+                let _: JoinHandle<Result<()>> = tokio::spawn(async move {
+                    let mut stream = input.execute(i).await?;
+                    let mut counter = 0;
+                    while let Some(result) = stream.next().await {
+                        match partitioning {
+                            Partitioning::RoundRobinBatch(_) => {
```

Comment:
The hash partition is not yet implemented here?

Reply:
No. I filed https://issues.apache.org/jira/browse/ARROW-11011 to implementing hash partitioning as a separate PR since it will be quite a lot of work.

`RepartitionExec::try_new` returns a `DataFusionError::NotImplemented` error if you try and create it with the hash partitioning scheme.

Reply:
Ok 👍 makes sense!